pkg/deploy/reconcile_manager.go

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package deploy
+
+import (
+	"reflect"
+	"runtime"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Reconcilable interface {
+	// Reconcile object.
+	Reconcile(ctx *DeployContext) (result reconcile.Result, done bool, err error)
+	// Does finalization (removes cluster scope objects, etc)
+	Finalize(ctx *DeployContext) (done bool, err error)
+	// Does registration
+	Register(rm *ReconcileManager)
+}
+
+type ReconcileManager struct {
+	reconcilers      []Reconcilable
+	failedReconciler Reconcilable
+}
+
+func NewReconcileManager() *ReconcileManager {
+	return &ReconcileManager{
+		reconcilers:      make([]Reconcilable, 0),
+		failedReconciler: nil,
+	}
+}
+
+func (rm *ReconcileManager) RegisterReconciler(reconciler Reconcilable) {
+	rm.reconcilers = append(rm.reconcilers, reconciler)
+}
+
+// Reconcile all objects in a order they have been added
+// If reconciliation failed then CheCluster status will be updated accordingly.
+func (rm *ReconcileManager) ReconcileAll(ctx *DeployContext) (reconcile.Result, bool, error) {
+	for _, reconciler := range rm.reconcilers {
+		result, done, err := reconciler.Reconcile(ctx)
+		if err != nil {
+			rm.failedReconciler = reconciler
+			if err := SetStatusDetails(ctx, InstallOrUpdateFailed, err.Error(), ""); err != nil {
```

Comment:
Shouldn't we also log the error?

Reply:
No, since runtime controller print it itself.